Make FontCascade loggable

smfr opened the pull request:
#### 1758ef2b35fc15e6f4e94e75509c325511701510
<pre>
Make FontCascade loggable
<a href="https://bugs.webkit.org/show_bug.cgi?id=283932">https://bugs.webkit.org/show_bug.cgi?id=283932</a>
<a href="https://rdar.apple.com/140805857">rdar://140805857</a>

Reviewed by Tim Nguyen.

FontCascade changes are responsible for a lot of Layout style diffs.
It&apos;s useful to be able to get some logging information about FontCascade.

This is not complete logging, just enough to give some basic information,
and reflect why operator== can return false.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::fontSelector const):
* Source/WebCore/platform/graphics/FontSelector.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontSelector.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::logStyleDifference): Just tidy it up a bit.
(WebCore::RenderElement::setStyle):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::dumpDifferences const):

Canonical link: <a href="https://commits.webkit.org/287245@main">https://commits.webkit.org/287245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c0929044c6d3e7811b2f4c28e31b139cba358e9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83575 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61795 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19720 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51834 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71815 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42099 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49184 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26014 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28520 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84954 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6276 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70027 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69278 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17257 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13326 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12064 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6225 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6192 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9639 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7983 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
```diff
@@ -158,4 +159,28 @@ void FontCascadeDescription::resolveFontSizeAdjustFromFontIfNeeded(const Font& f
+            ts << ", ";
+        ts << familiy;
+        first = false;
+    }
```
interleave is a nice pattern for this.

```
ts << interleave(fontCascadeDescription.families(), [&](auto& family) {
    return family;
}, ", "_s);
```

nt1m commented on Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
```diff
@@ -158,4 +159,28 @@ void FontCascadeDescription::resolveFontSizeAdjustFromFontIfNeeded(const Font& f
+    for (auto& familiy : fontCascadeDescription.families()) {
+        if (!first)
+            ts << ", ";
+        ts << familiy;
```
typo: familiy

nt1m reviewed (approved):
Please use ""_s through out for strings.

smfr committed:
Make FontCascade loggable
https://bugs.webkit.org/show_bug.cgi?id=283932
rdar://140805857

Reviewed by Tim Nguyen.

FontCascade changes are responsible for a lot of Layout style diffs.
It's useful to be able to get some logging information about FontCascade.

This is not complete logging, just enough to give some basic information,
and reflect why operator== can return false.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::fontSelector const):
* Source/WebCore/platform/graphics/FontSelector.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontSelector.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::logStyleDifference): Just tidy it up a bit.
(WebCore::RenderElement::setStyle):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::dumpDifferences const):

Canonical link: https://commits.webkit.org/287245@main
